noxfile.py

```diff
@@ -36,28 +75,106 @@ def run_readable(session, mode="fmt"):
         "ghcr.io/bobheadxi/readable:v0.5.0@sha256:423c133e7e9ca0ac20b0ab298bd5dbfa3df09b515b34cbfbbe8944310cc8d9c9",
         mode,
         "![.]**/*.md",
+        external=True,
     )
 
 
+def run_shellcheck(session, mode="check"):
+    shellcheck_cmd = [
+        "docker",
+        "run",
+        "--platform",
+        "linux/amd64",  # while this image is multi-arch, we cannot use digest with multi-arch images
+        "--rm",
+        "-v",
+        f"{ROOT.absolute()}:/mnt",
+        "-w",
+        "/mnt",
+        "-q",
+        "koalaman/shellcheck:0.9.0@sha256:a527e2077f11f28c1c1ad1dc784b5bc966baeb3e34ef304a0ffa72699b01ad9c",
+    ]
+
+    files = list_files(suffix=".sh")
+    if not files:
+        session.log("No shell files found")
+        return
+    shellcheck_cmd.extend(files)
+
+    if mode == "fmt":
+        with tempfile.NamedTemporaryFile(mode="w+") as diff_file:
+            session.run(
+                *shellcheck_cmd,
+                "--format=diff",
+                external=True,
+                stdout=diff_file,
+                success_codes=[0, 1],
+            )
+            diff_file.seek(0)
+            diff = diff_file.read()
+            if len(diff.splitlines()) > 1:  # ignore single-line message
+                session.log("Applying shellcheck patch:\n%s", diff)
+                subprocess.run(
+                    ["patch", "-p1"],
+                    input=diff,
+                    text=True,
+                    check=True,
+                )
+
+    session.run(*shellcheck_cmd, external=True)
+
+
 @nox.session(name="format", python=PYTHON_DEFAULT_VERSION)
 def format_(session):
-    session.run_always("pdm", "install", "--dev")
-    session.run("black", ".")
+    """Lint the code and apply fixes in-place whenever possible."""
+    install(session, "lint")
```

Comment:
suggestion: installing default group on lint sessions can extend CI time and also use a lot of disk space locally (eg.  when we develop project using CUDA) - my suggestion is to add `-no-default` for lint sessions - like I did in https://github.com/backend-developers-ltd/compute-horde-facilitator-sdk/blob/master/noxfile.py :)

Reply:
sounds good